File: go/quick.go
package main

import (
	"math/rand"
)

// quick implements quick sort algorithm. The function takes an array and its length.
func quick(a []int) {
	if len(a) <= 1 {
		return
	}
	r := rand.Intn(len(a))
	a[0], a[r] = a[r], a[0]
	k := 0
	for i := 1; i < len(a); i++ {
		if a[i] < a[0] {
			k++
			a[k], a[i] = a[i], a[k]
		}
	}
	a[0], a[k] = a[k], a[0]
	quick(a[:k])
	quick(a[k+1:])
}

File: python/quick.py
import random
import util

# sort implements quick sort algorithm. The function takes an array.
def sort(a):
	_sort(a, 0, len(a))

def _sort(a, l, n):
	if n <= 1:
		return
	util.swap(a, l, l+random.randint(0, n-1))
	k = l
	for i in range(l+1, l+n):
		if a[i] < a[l]:
			util.swap(a, ++k, i)
	util.swap(a, l, k)
	_sort(a, l, k-l)
	_sort(a, k+1, l+n-k-1)
